### PR TITLE
Add ImportSignups working with file (*.txt) and text

### PR DESCRIPTION
**Pull request by 3Mydlo3**
TODO:
- [x] Description
- [x] Maybe some cleanup

### CODE REVIEW

File: ArmaforcesMissionBot/Exceptions/NotAuthorizedException.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
+﻿using System;
+
+namespace ArmaforcesMissionBot.Exceptions {
+    public class NotAuthorizedException : Exception {
+        private const string DefaultMessage = "User is not authorized to use to perform this action!";
```
Comment (nielu) [inline ArmaforcesMissionBot/Exceptions/NotAuthorizedException.cs]: Typo? "User is not authorized to perform this action"

File: ArmaforcesMissionBot/Features/Signups/Importer/SignupImporter.cs [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+
+namespace ArmaforcesMissionBot.Features.Signups.Importer {
+    public class SignupImporter {
+        private ICommandContext Context { get; }
+        private CommandService CommandService { get; }
+        private IServiceProvider ServiceProvider { get; }
+        private IModule Module { get; }
+
+        public SignupImporter
+        (
```
Comment (nielu) [inline ArmaforcesMissionBot/Features/Signups/Importer/SignupImporter.cs]: image

File: ArmaforcesMissionBot/Features/Signups/Importer/SignupImporter.cs [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+        public async Task ProcessMessage(string message) {
+            var lineBreak = GetLineEnding(message);
+            var lines = ReadLines(message, lineBreak);
+            var commands = ParseCommands(lines, lineBreak);
+
+            if (!commands.Any())
+                await Module.ReplyWithException<InvalidCommandParametersException>("Nie udało się odczytać komend.");
+
+            foreach (var command in commands) await ProcessCommand(command);
```
Comment (nielu) [inline ArmaforcesMissionBot/Features/Signups/Importer/SignupImporter.cs:39]: Wouldn't something like this 
```cs
var tasks = commands.Select(c => ProcessCommand(c);
await Tasks.WhenAll(tasks);
```
perform better? 
Ref <https://stackoverflow.com/a/18668733/7411472>

Comment (3Mydlo3) [reply ArmaforcesMissionBot/Features/Signups/Importer/SignupImporter.cs:39]: Commands must be executed in correct order (at least the first one which starts the signups creation procedure, not sure about the rest). I'll take a look into signups creation 🍝.

File: ArmaforcesMissionBot.Tests/Features/Signups/Importer/SignupImporterTests.cs [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        }
+
+        [Theory]
+        [InlineData(new[] {"dsdsd", "AF!hsda", "#dsds", "asdas"}, new[] { "hsda\nasdas" })]
+        [InlineData(new[] {"AF!dsdsd", "AF!hsda", "#dsds", "asdas"}, new[] { "dsdsd", "hsda\nasdas" })]
+        [InlineData(new[] {"AF!dsdsd", "hsda", "#AF!dsds", "asdas"}, new[] { "dsdsd\nhsda\nasdas" })]
+        [InlineData(new[] {"AF!dsdsd", "", "asdas"}, new[] { "dsdsd\n\nasdas" })]
+        [InlineData(new[] {"dsdsd", "hsda", "#AF!dsds", "asdas"}, new string[0])]
+        [InlineData(new[] {"dsdsd", "hsda", "//AF!dsds", "asdas"}, new string[0])]
+        public void ParseCommands(IEnumerable<string> lines, IEnumerable<string> expectedCommandsList) {
```
Comment (nielu) [inline ArmaforcesMissionBot.Tests/Features/Signups/Importer/SignupImporterTests.cs:41]: I'd include also a test for AF! moniker in the middle of a word, eg. "dddAF!dupa"

File: ArmaforcesMissionBot/Features/Signups/Importer/SignupImporter.cs [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+            return message.Split(lineEnding)
+                .Select(x => x + lineEnding);
+        }
+
+        /// <summary>
+        ///     Checks and returns used line endings in given string.
+        /// </summary>
+        /// <param name="message">Message to check</param>
+        /// <returns>"\r\n" or "\n"</returns>
+        public static string GetLineEnding(string message)
```
Comment (nielu) [inline ArmaforcesMissionBot/Features/Signups/Importer/SignupImporter.cs:177]: Wouldn't it be easier to simply run .replace("\r\n", "\n") on every message?

File: ArmaforcesMissionBot/Modules/Signups.cs [hunk 6]
```diff
@@ -1,38 +1,73 @@
...
+        [Command("importuj-zapisy")]
+        [Summary("Importuje zapisy z załączonego pliku *.txt. lub z wiadomości (preferując plik txt jeżeli obie rzeczy są). " +
+                 "Czyta plik/wiadomość linia po linii, dołączając linie bez prefixu 'AF!' do poprzedniej komendy " +
+                 "a następnie wywołuje komendy w kolejności. " +
+                 "Ignoruje linie zaczynające się od '#' oraz '//' umożliwiając komentarze.")]
+        [ContextDMOrChannel]
+        public async Task ImportSignups([Remainder]string missionContent = null) {
+            if (_client.GetGuild(_config.AFGuild)
+                .GetUser(Context.User.Id)
+                .Roles.All(x => x.Id != _config.MissionMakerRole))
```
Comment (nielu) [inline ArmaforcesMissionBot/Modules/Signups.cs:45]: Shouldn't this be Roles.Any?

**Review by nielu — APPROVED** [review]

File: ArmaforcesMissionBot/Modules/Signups.cs (hunk 6 above)
Comment (3Mydlo3) [reply ArmaforcesMissionBot/Modules/Signups.cs:45]: It is ok, we need to check whether any user role is the mission maker role, and if not then return him an error. It's just inverted for less nesting.

**Pull request merged**